core/src/lib/channel.h

```diff
@@ -0,0 +1,349 @@
+// interact with either the input or the output.
+// This ensures that there is only one producer (who writes to the input)
+// and one consumer (who reads from the output).
+
+struct failed_to_acquire_lock {};
+struct channel_closed {};
+
+template <typename T> class queue {
+  struct internal {
+    std::vector<T> data;
+    bool in_dead;
+    bool out_dead;
+  };
+
+  synchronized<internal> shared{};
+  std::condition_variable in_update{};
+  std::condition_variable out_update{};
+  std::size_t max_size;
+
+ public:
+  queue(std::size_t max_size) : max_size(max_size) {}
+  queue(const queue&) = delete;
+  queue& operator=(const queue&) = delete;
+  queue(queue&&) = delete;
+  queue& operator=(queue&&) = delete;
+
+  using locked_type = decltype(shared.lock());
+
+  class handle {
+    locked_type locked;
+    std::condition_variable* update;
+
+   public:
+    handle(locked_type locked, std::condition_variable* update)
+        : locked{std::move(locked)}, update(update)
+    {
+    }
+    handle(const handle&) = delete;
+    handle& operator=(const handle&) = delete;
+    handle(handle&&) = delete;
+    handle& operator=(handle&&) = delete;
+
+    std::vector<T>& data() { return locked->data; }
+
+    ~handle() { update->notify_one(); }
+  };
+
+  /* the *_lock functions return std::nullopt only if the channel is closed
+   * otherwise they wait until they get the lock
+   * The try_*_lock functions instead return a tristate indicating whether
+   * they succeeded, failed to acquire the lock or if the channel was closed
+   * from the other side. */
+
+
+  std::optional<handle> read_lock()
+  {
+    auto locked = shared.lock();
+    if (locked->out_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to read from channel that was closed from the read side.\n");
+      return std::nullopt;
+    }
+
+    locked.wait(in_update, [](const auto& intern) {
+      return intern.data.size() > 0 || intern.in_dead;
+    });
+    if (locked->data.size() == 0) {
+      return std::nullopt;
+    } else {
+      return std::make_optional<handle>(std::move(locked), &out_update);
+    }
+  }
+
+  std::optional<handle> write_lock()
+  {
+    auto locked = shared.lock();
+    locked.wait(out_update, [max_size = max_size](const auto& intern) {
+      return intern.data.size() < max_size || intern.out_dead;
+    });
+    if (locked->in_dead) {
+      // note(ssura): This happening is programmer error.
+      // Maybe we should assert this instead ?
+      Dmsg0(50,
+            "Tried to write to channel that was closed from the write side.\n");
+      return std::nullopt;
+    }
+    if (locked->out_dead) {
+      return std::nullopt;
+    } else {
+      return std::make_optional<handle>(std::move(locked), &in_update);
+    }
+  }
+
+  using try_result
+      = std::variant<handle, failed_to_acquire_lock, channel_closed>;
```

Comment:
I'm not 100% convinced that `std::variant<>` is the best type for this.
We may want to get the library that C++23's `std::expected` is based on and use that in conjunction with an enum for the errors.

This would allow us to replace `std::optional<>` and unify the return values of `*_lock()` and `try_*lock()` to `expected<handle, error_enum>` opening the door for even more refactoring. Mimicing the standard libary's locking, we could replace the `try_*_lock()` interface with an `*_lock(try_to_lock_t)` overload.

Reply:
i think you are right.  Ill have to think about a way to do this.